### PR TITLE
KYLO-250 Added functionality for registering S3 backed tables to RegisterFeedTables nifi processor.

### PR DESCRIPTION
**Pull request by GregP86**
Added three properties to RegisterFeedTables processor. One that allows the user to select if they would like their tables to be backed by S3 or HDFS called "Storage Type". Another property that allows the user to specify an S3 bucket if necessary entitled "S3 Bucket". Lastly, a property that allows for the user to select which protocol they wish to use for accessing S3, either s3n or s3a, this property is called "S3 Protocol". Added logic for creating a path based on these properties to TableType.deriveLocationSpecification.

### CODE REVIEW

File: integrations/nifi/nifi-nar-bundles/nifi-core-bundle/nifi-core-processors/src/main/java/com/thinkbiganalytics/nifi/v2/ingest/RegisterFeedTables.java [hunk 1]
```diff
@@ -109,6 +110,39 @@
...
+    private static String S3N = "s3n";
+    private static String S3A = "s3a";
+    /**
+     * Property indicating which protocol is to be used if using s3
+     */
+    public static final PropertyDescriptor S3_PROTOCOL = new PropertyDescriptor.Builder()
+            .name("S3 Protocol")
+            .description("Specifies which S3 protocol is to be used if using S3 storage.")
+            .required(false)
+            .allowableValues(S3N, S3A)
```
Comment (harschware) [inline integrations/nifi/nifi-nar-bundles/nifi-core-bundle/nifi-core-processors/src/main/java/com/thinkbiganalytics/nifi/v2/ingest/RegisterFeedTables.java]: I'm wondering if S3N and S3A are the only and correct choices here?   I read a little on S3 support in Hadoop 2.7 (https://wiki.apache.org/hadoop/AmazonS3) and it strongly encourages s3a for Hadoop2.7+. Can Kylo try to assert 2.7+ going forward?   It also states S3 will not be supported in Hadoop3.0 so it seems correct that is not a choice here.

According to this (http://docs.aws.amazon.com/emr/latest/ManagementGuide/emr-plan-file-systems.html) S3N is strongly discouraged.

I'm wondering if s3n should be removed..

**Comment by freastro** [comment]
It looks like Hive just needs a URI for the Location. What if we allowed the Feed Root Path, Profile Root Path, and the Master Root Path to be URIs? I think that would support additional FileSystems (besides hdfs, s3n, and s3a) and it would be a simpler change.

**Comment by scottreisdorf** [comment]
Based upon Greg Harts suggestion I think you might be good by changing the following classes:


1. TableRegisterConfiguration
     - Allow for the feedRoot, profileRoot, masterRoot to be URIs
     - change the pathForTableType(TableType type)  method to make it work with URIs
2. TableType
   - Ensure the TableType#deriveLocationSpecification() method can work with the URI change made in Step 1
3. TableRegisterSupport is the one that calls the items in 1,2 above
      -TableRegisterSupport#createDDL   - line 187
           -  String locationSQL = tableType.deriveLocationSpecification(config.pathForTableType(tableType), source, entity);

**Comment by GregP86** [comment]
We removed the added properties from the RegisterFeedTable processor as requested. We attempted to use URI objects for the root paths but we ran into issues when concatenating the URIs.   Using Path classes instead seemed to handle this better than URI. We also attempted to use Strings but this resulted in some messy root paths such as missing or duplicated forward slashes. The Path class handles these situations cleanly. In order to use the Path object with different filesystems we needed to remove the ".toAbsolutePath()" call. We do not think this will cause an issue but please advise if this method call is needed.

File: integrations/nifi/nifi-nar-bundles/nifi-core-bundle/nifi-core-processors/src/main/java/com/thinkbiganalytics/util/TableType.java [hunk 2]
```diff
@@ -80,10 +80,13 @@ public String deriveLocationSpecification(Path tableLocation, String source, Str
         Validate.notNull(entity, "entity expected");
 
         Path path = tableLocation.resolve(source).resolve(entity).resolve(tableSuffix);
+        String location = path.toString().replace(":/","://");
```
Comment (freastro) [inline integrations/nifi/nifi-nar-bundles/nifi-core-bundle/nifi-core-processors/src/main/java/com/thinkbiganalytics/util/TableType.java]: What paths would require replacing `:/` with `://`?

File: integrations/nifi/nifi-nar-bundles/nifi-core-bundle/nifi-core-processors/src/test/java/com/thinkbiganalytics/ingest/TableRegisterSupportTest.java [hunk 3]
```diff
@@ -120,6 +122,26 @@ public void testTableCreate() {
...
+        ColumnSpec[] parts = ColumnSpec.createFromString("year|int\ncountry|string");
+        TableRegisterConfiguration conf = new TableRegisterConfiguration("s3a://testBucket/model.db/","s3a://testBucket/model.db/","s3a://testBucket/app/warehouse/");
+        TableRegisterSupport support = new TableRegisterSupport(connection, conf);
+        TableType[] tableTypes = new TableType[]{TableType.FEED, TableType.INVALID, TableType.VALID, TableType.MASTER};
+        for (TableType tableType : tableTypes) {
+            String ddl =
+                    support.createDDL("bar", "employee", specs, parts, "ROW FORMAT SERDE 'org.apache.hadoop.hive.serde2.OpenCSVSerde'", "stored as orc", "tblproperties (\"orc.compress\"=\"SNAPPY\")",
+                            tableType);
+            String location = StringUtils.substringBetween(ddl, "LOCATION '", "'");
+            if(tableType.toString().equalsIgnoreCase("master")){
```
Comment (freastro) [inline integrations/nifi/nifi-nar-bundles/nifi-core-bundle/nifi-core-processors/src/test/java/com/thinkbiganalytics/ingest/TableRegisterSupportTest.java]: Enums can be compared using the == operator, for example: `tableType == TypeType.MASTER`.

File: integrations/nifi/nifi-nar-bundles/nifi-core-bundle/nifi-core-processors/src/main/java/com/thinkbiganalytics/util/TableType.java (hunk 2 above)
Comment (rashidaligee) [inline integrations/nifi/nifi-nar-bundles/nifi-core-bundle/nifi-core-processors/src/main/java/com/thinkbiganalytics/util/TableType.java]: If a path contains // slashes, Paths APIs take //  fixed the path by removing one slash. But in URIs like `s3://bucket-name` we do require double forward slashes. Paths objets handles many other scenarios but lack a forward slash. After path is calculated, extra forward slash is added to fix the URI path.

File: integrations/nifi/nifi-nar-bundles/nifi-core-bundle/nifi-core-processors/src/test/java/com/thinkbiganalytics/ingest/TableRegisterSupportTest.java (hunk 3 above)
Comment (rashidaligee) [inline integrations/nifi/nifi-nar-bundles/nifi-core-bundle/nifi-core-processors/src/test/java/com/thinkbiganalytics/ingest/TableRegisterSupportTest.java]: +1

**Comment by freastro** [comment]
@GregP86, could you reformat the code using our [development guidelines](http://kylo.readthedocs.io/en/latest/developer-guides/ContributingtoKylo.html#development-guidelines)? It's just a simple Eclipse or IntelliJ code style that you import and run. Thanks!

**Comment by harschware** [comment]
@GregP86  can you please share the details of your test environment?    I get the following error:

```
2017-03-21 14:21:57,093 INFO [Timer-Driven Process Thread-5] c.t.nifi.v2.hdfs.HDFSSupport Creating path s3://thinkbig.tim/PR6/model.db/demo55 
2017-03-21 14:21:57,094 ERROR [Timer-Driven Process Thread-5] c.t.nifi.v2.hdfs.HDFSSupport Failed to create path s3://thinkbig.tim/PR6/model.db/demo55 due to {}
java.lang.IllegalArgumentException: Wrong FS: s3://thinkbig.tim/PR6/model.db/demo55, expected: hdfs://ip-172-31-25-248.us-west-2.compute.internal:8020
	at org.apache.hadoop.fs.FileSystem.checkPath(FileSystem.java:645) ~[hadoop-common-2.6.2.jar:na]
	at org.apache.hadoop.hdfs.DistributedFileSystem.getPathName(DistributedFileSystem.java:193) ~[hadoop-hdfs-2.6.2.jar:na]
	at org.apache.hadoop.hdfs.DistributedFileSystem.access$000(DistributedFileSystem.java:105) ~[hadoop-hdfs-2.6.2.jar:na]
	at org.apache.hadoop.hdfs.DistributedFileSystem$18.doCall(DistributedFileSystem.java:1118) ~[hadoop-hdfs-2.6.2.jar:na]
	at org.apache.hadoop.hdfs.DistributedFileSystem$18.doCall(DistributedFileSystem.java:1114) ~[hadoop-hdfs-2.6.2.jar:na]
	at org.apache.hadoop.fs.FileSystemLinkResolver.resolve(FileSystemLinkResolver.java:81) ~[hadoop-common-2.6.2.jar:na]
	at org.apache.hadoop.hdfs.DistributedFileSystem.getFileStatus(DistributedFileSystem.java:1114) ~[hadoop-hdfs-2.6.2.jar:na]
	at com.thinkbiganalytics.nifi.v2.hdfs.HDFSSupport.createFolder(HDFSSupport.java:58) ~[kylo-nifi-hadoop-processors-0.8.0-SNAPSHOT.jar:na]
	at com.thinkbiganalytics.nifi.v2.hdfs.CreateHDFSFolder.onTrigger(CreateHDFSFolder.java:211) [kylo-nifi-hadoop-processors-0.8.0-SNAPSHOT.jar:na]
	at org.apache.nifi.processor.AbstractProcessor.onTrigger(AbstractProcessor.java:27) [nifi-api-1.0.0.jar:1.0.0]
	at org.apache.nifi.controller.StandardProcessorNode.onTrigger(StandardProcessorNode.java:1064) [nifi-framework-core-1.0.0.jar:1.0.0]
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:136) [nifi-framework-core-1.0.0.jar:1.0.0]
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:47) [nifi-framework-core-1.0.0.jar:1.0.0]
	at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:132) [nifi-framework-core-1.0.0.jar:1.0.0]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_121]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [na:1.8.0_121]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [na:1.8.0_121]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [na:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_121]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_121]
2017-03-21 14:21:57,094 ERROR [Timer-Driven Process Thread-5] c.t.nifi.v2.hdfs.CreateHDFSFolder CreateHDFSFolder[id=2b1f1ef0-4e0a-1bd6-ae68-e9052773e3ac] failed folder creation java.lang.RuntimeException: Failed to create path [s3://thinkbig.tim/PR6/model.db/demo55]: java.lang.RuntimeException: Failed to create path [s3://thinkbig.tim/PR6/model.db/demo55]
2017-03-21 14:21:57,095 ERROR [Timer-Driven Process Thread-5] c.t.nifi.v2.hdfs.CreateHDFSFolder 
java.lang.RuntimeException: Failed to create path [s3://thinkbig.tim/PR6/model.db/demo55]
	at com.thinkbiganalytics.nifi.v2.hdfs.HDFSSupport.createFolder(HDFSSupport.java:70) ~[kylo-nifi-hadoop-processors-0.8.0-SNAPSHOT.jar:na]
	at com.thinkbiganalytics.nifi.v2.hdfs.CreateHDFSFolder.onTrigger(CreateHDFSFolder.java:211) ~[kylo-nifi-hadoop-processors-0.8.0-SNAPSHOT.jar:na]
	at org.apache.nifi.processor.AbstractProcessor.onTrigger(AbstractProcessor.java:27) [nifi-api-1.0.0.jar:1.0.0]
	at org.apache.nifi.controller.StandardProcessorNode.onTrigger(StandardProcessorNode.java:1064) [nifi-framework-core-1.0.0.jar:1.0.0]
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:136) [nifi-framework-core-1.0.0.jar:1.0.0]
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:47) [nifi-framework-core-1.0.0.jar:1.0.0]
	at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:132) [nifi-framework-core-1.0.0.jar:1.0.0]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_121]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [na:1.8.0_121]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [na:1.8.0_121]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [na:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_121]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_121]
Caused by: java.lang.IllegalArgumentException: Wrong FS: s3://thinkbig.tim/PR6/model.db/demo55, expected: hdfs://ip-172-31-25-248.us-west-2.compute.internal:8020
	at org.apache.hadoop.fs.FileSystem.checkPath(FileSystem.java:645) ~[hadoop-common-2.6.2.jar:na]
	at org.apache.hadoop.hdfs.DistributedFileSystem.getPathName(DistributedFileSystem.java:193) ~[hadoop-hdfs-2.6.2.jar:na]
	at org.apache.hadoop.hdfs.DistributedFileSystem.access$000(DistributedFileSystem.java:105) ~[hadoop-hdfs-2.6.2.jar:na]
	at org.apache.hadoop.hdfs.DistributedFileSystem$18.doCall(DistributedFileSystem.java:1118) ~[hadoop-hdfs-2.6.2.jar:na]
	at org.apache.hadoop.hdfs.DistributedFileSystem$18.doCall(DistributedFileSystem.java:1114) ~[hadoop-hdfs-2.6.2.jar:na]
	at org.apache.hadoop.fs.FileSystemLinkResolver.resolve(FileSystemLinkResolver.java:81) ~[hadoop-common-2.6.2.jar:na]
	at org.apache.hadoop.hdfs.DistributedFileSystem.getFileStatus(DistributedFileSystem.java:1114) ~[hadoop-hdfs-2.6.2.jar:na]
	at com.thinkbiganalytics.nifi.v2.hdfs.HDFSSupport.createFolder(HDFSSupport.java:58) ~[kylo-nifi-hadoop-processors-0.8.0-SNAPSHOT.jar:na]
	... 13 common frames omitted
```

I set the following in my application.properties config:
```
# Root HDFS location for Hive ingest processing tables (raw,valid,invalid)
config.hive.ingest.root=s3://thinkbig.tim/PR6/model.db
# Root HDFS location for Hive profile table
config.hive.profile.root=s3://thinkbig.tim/PR6/model.db
# Root HDFS location for Hive master table
config.hive.master.root=s3://thinkbig.tim/PR6/app/warehouse
```

**Comment by GregP86** [comment]
I was just looking at that S3 error. If I am not mistaken it is one that was recently encountered by @smtannehill as well. He said that if an S3 bucket was completely empty he got this error but as soon as any file was present the error went away.

**Pull request merged**